File: backend/app/schemas/monitoring.py
"""
Pydantic schemas for system monitoring API.

Provides request/response models for:
- CPU, Memory, Network, Disk I/O metrics
- Process tracking
- Retention configuration
"""

from datetime import datetime
from typing import Dict, List, Optional
from enum import Enum

from pydantic import BaseModel, Field


# ===== Metric Type Enum =====

class MetricTypeEnum(str, Enum):
    """Types of metrics."""
    CPU = "cpu"
    MEMORY = "memory"
    NETWORK = "network"
    DISK_IO = "disk_io"
    PROCESS = "process"


class DataSource(str, Enum):
    """Data source for history queries."""
    AUTO = "auto"
    MEMORY = "memory"
    DATABASE = "database"


# ===== Sample Schemas =====

class CpuSampleSchema(BaseModel):
    """CPU metrics sample."""
    timestamp: datetime
    usage_percent: float
    frequency_mhz: Optional[float] = None
    temperature_celsius: Optional[float] = None
    core_count: Optional[int] = None
    thread_count: Optional[int] = None
    p_core_count: Optional[int] = None  # Intel Performance cores
    e_core_count: Optional[int] = None  # Intel Efficiency cores

    class Config:
        from_attributes = True


class MemorySampleSchema(BaseModel):
    """Memory metrics sample."""
    timestamp: datetime
    used_bytes: int
    total_bytes: int
    percent: float
    available_bytes: Optional[int] = None
    baluhost_memory_bytes: Optional[int] = None  # Memory used by BaluHost processes

    class Config:
        from_attributes = True


class NetworkSampleSchema(BaseModel):
    """Network metrics sample."""
    timestamp: datetime
    download_mbps: float
    upload_mbps: float
    bytes_sent: Optional[int] = None
    bytes_received: Optional[int] = None

    class Config:
        from_attributes = True


class DiskIoSampleSchema(BaseModel):
    """Disk I/O metrics sample."""
    timestamp: datetime
    disk_name: str
    read_mbps: float
    write_mbps: float
    read_iops: float
    write_iops: float
    avg_response_ms: Optional[float] = None
    active_time_percent: Optional[float] = None

    class Config:
        from_attributes = True


class ProcessSampleSchema(BaseModel):
    """Process metrics sample."""
    timestamp: datetime
    process_name: str
    pid: int
    cpu_percent: float
    memory_mb: float
    status: str
    is_alive: bool = True

    class Config:
        from_attributes = True


# ===== Request Schemas =====

class MetricHistoryRequest(BaseModel):
    """Request parameters for metric history."""
    start: Optional[datetime] = None
    end: Optional[datetime] = None
    source: DataSource = DataSource.AUTO
    limit: int = Field(default=1000, ge=1, le=10000)


class TimeRangeEnum(str, Enum):
    """Predefined time ranges."""
    TEN_MINUTES = "10m"
    ONE_HOUR = "1h"
    TWENTY_FOUR_HOURS = "24h"
    SEVEN_DAYS = "7d"


# ===== Response Schemas =====

class CurrentCpuResponse(BaseModel):
    """Current CPU metrics response."""
    timestamp: datetime
    usage_percent: float
    frequency_mhz: Optional[float] = None
    temperature_celsius: Optional[float] = None
    core_count: Optional[int] = None
    thread_count: Optional[int] = None
    p_core_count: Optional[int] = None  # Intel Performance cores
    e_core_count: Optional[int] = None  # Intel Efficiency cores


class CurrentMemoryResponse(BaseModel):
    """Current memory metrics response."""
    timestamp: datetime
    used_bytes: int
    total_bytes: int
    percent: float
    available_bytes: Optional[int] = None
    baluhost_memory_bytes: Optional[int] = None  # Memory used by BaluHost processes


class CurrentNetworkResponse(BaseModel):
    """Current network metrics response."""
    timestamp: datetime
    download_mbps: float
    upload_mbps: float


class CurrentDiskIoResponse(BaseModel):
    """Current disk I/O metrics response."""
    disks: Dict[str, Optional[DiskIoSampleSchema]]


class CurrentProcessResponse(BaseModel):
    """Current process status response."""
    processes: Dict[str, Optional[ProcessSampleSchema]]


class MetricHistoryResponse(BaseModel):
    """Generic metric history response."""
    metric_type: str
    samples: List
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None
    sample_count: int
    source: str  # "memory" or "database"


class CpuHistoryResponse(BaseModel):
    """CPU history response."""
    samples: List[CpuSampleSchema]
    sample_count: int
    source: str


class MemoryHistoryResponse(BaseModel):
    """Memory history response."""
    samples: List[MemorySampleSchema]
    sample_count: int
    source: str


class NetworkHistoryResponse(BaseModel):
    """Network history response."""
    samples: List[NetworkSampleSchema]
    sample_count: int
    source: str


class DiskIoHistoryResponse(BaseModel):
    """Disk I/O history response."""
    disks: Dict[str, List[DiskIoSampleSchema]]
    available_disks: List[str]
    sample_count: int
    source: str


class ProcessHistoryResponse(BaseModel):
    """Process history response."""
    processes: Dict[str, List[ProcessSampleSchema]]
    sample_count: int
    source: str
    crashes_detected: int = 0


# ===== Retention Configuration =====

class RetentionConfigResponse(BaseModel):
    """Retention configuration for a metric type."""
    metric_type: str
    retention_hours: int
    db_persist_interval: int
    is_enabled: bool
    last_cleanup: Optional[datetime] = None
    samples_cleaned: int = 0


class RetentionConfigUpdate(BaseModel):
    """Update retention configuration."""
    retention_hours: int = Field(..., ge=1, le=8760, description="Retention period in hours (1-8760)")


class RetentionConfigListResponse(BaseModel):
    """List of all retention configurations."""
    configs: List[RetentionConfigResponse]


# ===== Database Stats =====

class MetricDatabaseStats(BaseModel):
    """Database statistics for a metric type."""
    metric_type: str
    count: int
    oldest: Optional[datetime] = None
    newest: Optional[datetime] = None
    retention_hours: int
    last_cleanup: Optional[datetime] = None
    total_cleaned: int = 0
    estimated_size_bytes: int = 0


class DatabaseStatsResponse(BaseModel):
    """Database statistics for all metrics."""
    metrics: Dict[str, MetricDatabaseStats]
    total_samples: int
    total_size_bytes: int


# ===== Monitoring Status =====

class MonitoringStatusResponse(BaseModel):
    """Overall monitoring status."""
    is_running: bool
    sample_count: int
    sample_interval: float
    buffer_size: int
    persist_interval: int
    last_cleanup: Optional[datetime] = None
    collectors: Dict[str, bool]

File: client/src/api/monitoring.ts
/**
 * API client for system monitoring
 *
 * Provides functions for:
 * - CPU, Memory, Network, Disk I/O current values and history
 * - Process tracking
 * - Retention configuration (admin only)
 */

import { apiClient } from '../lib/api';

// ===== Time Range Types =====

export type TimeRange = '10m' | '1h' | '24h' | '7d';
export type DataSource = 'auto' | 'memory' | 'database';

// ===== Sample Types =====

export interface CpuSample {
  timestamp: string;
  usage_percent: number;
  frequency_mhz?: number;
  temperature_celsius?: number;
  core_count?: number;
  thread_count?: number;
  p_core_count?: number;  // Intel Performance cores
  e_core_count?: number;  // Intel Efficiency cores
  thread_usages?: number[];  // Per-thread CPU usage percentages
}

export interface MemorySample {
  timestamp: string;
  used_bytes: number;
  total_bytes: number;
  percent: number;
  available_bytes?: number;
  baluhost_memory_bytes?: number;  // Memory used by BaluHost processes
}

export interface NetworkSample {
  timestamp: string;
  download_mbps: number;
  upload_mbps: number;
  bytes_sent?: number;
  bytes_received?: number;
}

export interface DiskIoSample {
  timestamp: string;
  disk_name: string;
  read_mbps: number;
  write_mbps: number;
  read_iops: number;
  write_iops: number;
  avg_response_ms?: number;
  active_time_percent?: number;
}

export interface ProcessSample {
  timestamp: string;
  process_name: string;
  pid: number;
  cpu_percent: number;
  memory_mb: number;
  status: string;
  is_alive: boolean;
}

// ===== Current Response Types =====

export interface CurrentCpuResponse {
  timestamp: string;
  usage_percent: number;
  frequency_mhz?: number;
  temperature_celsius?: number;
  core_count?: number;
  thread_count?: number;
  p_core_count?: number;  // Intel Performance cores
  e_core_count?: number;  // Intel Efficiency cores
  thread_usages?: number[];  // Per-thread CPU usage percentages
}

export interface CurrentMemoryResponse {
  timestamp: string;
  used_bytes: number;
  total_bytes: number;
  percent: number;
  available_bytes?: number;
  baluhost_memory_bytes?: number;  // Memory used by BaluHost processes
}

export type InterfaceType = 'wifi' | 'ethernet' | 'unknown';

export interface CurrentNetworkResponse {
  timestamp: string;
  download_mbps: number;
  upload_mbps: number;
  interface_type?: InterfaceType;
}

export interface CurrentDiskIoResponse {
  disks: Record<string, DiskIoSample | null>;
}

export interface CurrentProcessResponse {
  processes: Record<string, ProcessSample | null>;
}

// ===== History Response Types =====

export interface CpuHistoryResponse {
  samples: CpuSample[];
  sample_count: number;
  source: string;
}

export interface MemoryHistoryResponse {
  samples: MemorySample[];
  sample_count: number;
  source: string;
}

export interface NetworkHistoryResponse {
  samples: NetworkSample[];
  sample_count: number;
  source: string;
}

export interface DiskIoHistoryResponse {
  disks: Record<string, DiskIoSample[]>;
  available_disks: string[];
  sample_count: number;
  source: string;
}

export interface ProcessHistoryResponse {
  processes: Record<string, ProcessSample[]>;
  sample_count: number;
  source: string;
  crashes_detected: number;
}

// ===== Retention Config Types =====

export interface RetentionConfig {
  metric_type: string;
  retention_hours: number;
  db_persist_interval: number;
  is_enabled: boolean;
  last_cleanup?: string;
  samples_cleaned: number;
}

export interface RetentionConfigListResponse {
  configs: RetentionConfig[];
}

// ===== Database Stats Types =====

export interface MetricDatabaseStats {
  metric_type: string;
  count: number;
  oldest?: string;
  newest?: string;
  retention_hours: number;
  last_cleanup?: string;
  total_cleaned: number;
  estimated_size_bytes: number;
}

export interface DatabaseStatsResponse {
  metrics: Record<string, MetricDatabaseStats>;
  total_samples: number;
  total_size_bytes: number;
}

// ===== Monitoring Status Type =====

export interface MonitoringStatusResponse {
  is_running: boolean;
  sample_count: number;
  sample_interval: number;
  buffer_size: number;
  persist_interval: number;
  last_cleanup?: string;
  collectors: Record<string, boolean>;
}

// ===== API Functions =====

// CPU
export async function getCpuCurrent(): Promise<CurrentCpuResponse> {
  const response = await apiClient.get<CurrentCpuResponse>('/api/monitoring/cpu/current');
  return response.data;
}

export async function getCpuHistory(
  timeRange: TimeRange = '1h',
  source: DataSource = 'auto',
  limit: number = 1000
): Promise<CpuHistoryResponse> {
  const response = await apiClient.get<CpuHistoryResponse>('/api/monitoring/cpu/history', {
    params: { time_range: timeRange, source, limit },
  });
  return response.data;
}

// Memory
export async function getMemoryCurrent(): Promise<CurrentMemoryResponse> {
  const response = await apiClient.get<CurrentMemoryResponse>('/api/monitoring/memory/current');
  return response.data;
}

export async function getMemoryHistory(
  timeRange: TimeRange = '1h',
  source: DataSource = 'auto',
  limit: number = 1000
): Promise<MemoryHistoryResponse> {
  const response = await apiClient.get<MemoryHistoryResponse>('/api/monitoring/memory/history', {
    params: { time_range: timeRange, source, limit },
  });
  return response.data;
}

// Network
export async function getNetworkCurrent(): Promise<CurrentNetworkResponse> {
  const response = await apiClient.get<CurrentNetworkResponse>('/api/monitoring/network/current');
  return response.data;
}

export async function getNetworkHistory(
  timeRange: TimeRange = '1h',
  source: DataSource = 'auto',
  limit: number = 1000
): Promise<NetworkHistoryResponse> {
  const response = await apiClient.get<NetworkHistoryResponse>('/api/monitoring/network/history', {
    params: { time_range: timeRange, source, limit },
  });
  return response.data;
}

// Disk I/O
export async function getDiskIoCurrent(): Promise<CurrentDiskIoResponse> {
  const response = await apiClient.get<CurrentDiskIoResponse>('/api/monitoring/disk-io/current');
  return response.data;
}

export async function getDiskIoHistory(
  timeRange: TimeRange = '1h',
  source: DataSource = 'auto',
  diskName?: string,
  limit: number = 1000
): Promise<DiskIoHistoryResponse> {
  const params: Record<string, any> = { time_range: timeRange, source, limit };
  if (diskName) params.disk_name = diskName;
  const response = await apiClient.get<DiskIoHistoryResponse>('/api/monitoring/disk-io/history', {
    params,
  });
  return response.data;
}

// Processes
export async function getProcessesCurrent(): Promise<CurrentProcessResponse> {
  const response = await apiClient.get<CurrentProcessResponse>('/api/monitoring/processes/current');
  return response.data;
}

export async function getProcessesHistory(
  timeRange: TimeRange = '1h',
  source: DataSource = 'auto',
  processName?: string
): Promise<ProcessHistoryResponse> {
  const params: Record<string, any> = { time_range: timeRange, source };
  if (processName) params.process_name = processName;
  const response = await apiClient.get<ProcessHistoryResponse>('/api/monitoring/processes/history', {
    params,
  });
  return response.data;
}

// Retention Configuration (Admin only)
export async function getRetentionConfig(): Promise<RetentionConfigListResponse> {
  const response = await apiClient.get<RetentionConfigListResponse>('/api/monitoring/config/retention');
  return response.data;
}

export async function updateRetentionConfig(
  metricType: string,
  retentionHours: number
): Promise<RetentionConfig> {
  const response = await apiClient.put<RetentionConfig>(
    `/api/monitoring/config/retention/${metricType}`,
    { retention_hours: retentionHours }
  );
  return response.data;
}

// Database Stats (Admin only)
export async function getDatabaseStats(): Promise<DatabaseStatsResponse> {
  const response = await apiClient.get<DatabaseStatsResponse>('/api/monitoring/stats/database');
  return response.data;
}

// Monitoring Status
export async function getMonitoringStatus(): Promise<MonitoringStatusResponse> {
  const response = await apiClient.get<MonitoringStatusResponse>('/api/monitoring/status');
  return response.data;
}

// Manual Cleanup (Admin only)
export async function triggerCleanup(): Promise<{ message: string; deleted: Record<string, number>; total: number }> {
  const response = await apiClient.post<{ message: string; deleted: Record<string, number>; total: number }>(
    '/api/monitoring/cleanup'
  );
  return response.data;
}
